File: _examples/objective-c-uikit/TodoViewController.m
#import <UIKit/UIKit.h>

@interface TodoViewController : UIViewController <UITableViewDataSource, UITableViewDragDelegate, UITableViewDropDelegate>
@end

@implementation TodoViewController {
    UITextField *_textField;
    UIButton *_addButton;
    UITableView *_tableView;
    UISegmentedControl *_categorySegment;
    NSMutableArray<NSDictionary *> *_todos;
    NSArray<NSString *> *_categories;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _todos = [NSMutableArray array];
    _categories = @[@"work", @"daily"];
    [self setupUI];
}

- (void)setupUI {
    self.view.backgroundColor = [UIColor whiteColor];
    
    _categorySegment = [[UISegmentedControl alloc] initWithItems:_categories];
    _categorySegment.selectedSegmentIndex = 0;
    _categorySegment.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:_categorySegment];
    
    _textField = [[UITextField alloc] init];
    _textField.placeholder = @"Enter todo";
    _textField.borderStyle = UITextBorderStyleRoundedRect;
    _textField.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:_textField];
    
    _addButton = [UIButton buttonWithType:UIButtonTypeSystem];
    if (@available(iOS 13.0, *)) {
        UIImage *addIcon = [UIImage systemImageNamed:@"plus.circle.fill"];
        [_addButton setImage:addIcon forState:UIControlStateNormal];
    } else {
        [_addButton setTitle:@"Add" forState:UIControlStateNormal];
    }
    _addButton.translatesAutoresizingMaskIntoConstraints = NO;
    [_addButton addTarget:self action:@selector(addTodo) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_addButton];
    
    _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
    _tableView.translatesAutoresizingMaskIntoConstraints = NO;
    _tableView.dataSource = self;
    _tableView.dragDelegate = self;
    _tableView.dropDelegate = self;
    _tableView.dragInteractionEnabled = YES;
    [_tableView registerClass:[TodoCell class] forCellReuseIdentifier:@"TodoCell"];
    [self.view addSubview:_tableView];
    
    [NSLayoutConstraint activateConstraints:@[
        [_categorySegment.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:16],
        [_categorySegment.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:16],
        [_categorySegment.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-16],
        
        [_textField.topAnchor constraintEqualToAnchor:_categorySegment.bottomAnchor constant:16],
        [_textField.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:16],
        [_textField.trailingAnchor constraintEqualToAnchor:_addButton.leadingAnchor constant:-8],
        
        [_addButton.centerYAnchor constraintEqualToAnchor:_textField.centerYAnchor],
        [_addButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-16],
        [_addButton.widthAnchor constraintEqualToConstant:60],
        
        [_tableView.topAnchor constraintEqualToAnchor:_textField.bottomAnchor constant:16],
        [_tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [_tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [_tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor]
    ]];
}

- (void)addTodo {
    NSString *title = _textField.text;
    if (title.length == 0) return;
    NSString *category = _categories[_categorySegment.selectedSegmentIndex];
    [_todos addObject:@{@"title": title, @"category": category, @"isCompleted": @NO}];
    _textField.text = @"";
    [_tableView reloadData];
}

#pragma mark - UITableViewDragDelegate

- (NSArray<UIDragItem *> *)tableView:(UITableView *)tableView itemsForBeginningDragSession:(id<UIDragSession>)session atIndexPath:(NSIndexPath *)indexPath {
    NSDictionary *item = _todos[indexPath.row];
    NSItemProvider *itemProvider = [[NSItemProvider alloc] initWithObject:item[@"title"]];
    UIDragItem *dragItem = [[UIDragItem alloc] initWithItemProvider:itemProvider];
    dragItem.localObject = item;
    return @[dragItem];
}

#pragma mark - UITableViewDropDelegate

- (UITableViewDropProposal *)tableView:(UITableView *)tableView dropSessionDidUpdate:(id<UIDropSession>)session withDestinationIndexPath:(NSIndexPath *)destinationIndexPath {
    if (tableView.hasActiveDrag) {
        return [[UITableViewDropProposal alloc] initWithOperation:UIDropOperationMove intent:UITableViewDropIntentInsertAtDestinationIndexPath];
    }
    return [[UITableViewDropProposal alloc] initWithOperation:UIDropOperationForbidden];
}

- (void)tableView:(UITableView *)tableView performDropWithCoordinator:(id<UITableViewDropCoordinator>)coordinator {
    NSIndexPath *destinationIndexPath = coordinator.destinationIndexPath;
    if (!destinationIndexPath) return;
    
    id<UIDropItem> item = coordinator.items.firstObject;
    NSIndexPath *sourceIndexPath = item.sourceIndexPath;
    if (!sourceIndexPath) return;
    
    [tableView performBatchUpdates:^{
        NSDictionary *movedItem = [_todos[sourceIndexPath.row] mutableCopy];
        [_todos removeObjectAtIndex:sourceIndexPath.row];
        [_todos insertObject:movedItem atIndex:destinationIndexPath.row];
        [tableView moveRowAtIndexPath:sourceIndexPath toIndexPath:destinationIndexPath];
    } completion:nil];
    
    [coordinator dropItem:item.dragItem toRowAtIndexPath:destinationIndexPath];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _todos.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TodoCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TodoCell" forIndexPath:indexPath];
    NSDictionary *todo = _todos[indexPath.row];
    [cell configureWithTitle:todo[@"title"] category:todo[@"category"] isCompleted:[todo[@"isCompleted"] boolValue]];
    __weak typeof(self) weakSelf = self;
    cell.onCheckChanged = ^(BOOL isCompleted) {
        NSMutableDictionary *mutableTodo = [weakSelf.todos[indexPath.row] mutableCopy];
        mutableTodo[@"isCompleted"] = @(isCompleted);
        weakSelf.todos[indexPath.row] = mutableTodo;
    };
    cell.onArchive = ^{
        [weakSelf.todos removeObjectAtIndex:indexPath.row];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
    };
    return cell;
}

@end

#pragma mark - TodoCell

@interface TodoCell : UITableViewCell
@property (nonatomic, copy) void (^onCheckChanged)(BOOL);
@property (nonatomic, copy) void (^onArchive)(void);
@end

@implementation TodoCell {
    UISwitch *_checkBox;
    UILabel *_titleLabel;
    UILabel *_categoryLabel;
    UIButton *_archiveButton;
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setupUI {
    _checkBox = [[UISwitch alloc] init];
    _checkBox.translatesAutoresizingMaskIntoConstraints = NO;
    [_checkBox addTarget:self action:@selector(checkChanged) forControlEvents:UIControlEventValueChanged];
    [self.contentView addSubview:_checkBox];
    
    _titleLabel = [[UILabel alloc] init];
    _titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.contentView addSubview:_titleLabel];
    
    _categoryLabel = [[UILabel alloc] init];
    _categoryLabel.translatesAutoresizingMaskIntoConstraints = NO;
    _categoryLabel.textColor = [UIColor grayColor];
    [self.contentView addSubview:_categoryLabel];
    
    _archiveButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [_archiveButton setTitle:@"Archive" forState:UIControlStateNormal];
    _archiveButton.translatesAutoresizingMaskIntoConstraints = NO;
    [_archiveButton addTarget:self action:@selector(archiveTapped) forControlEvents:UIControlEventTouchUpInside];
    _archiveButton.hidden = YES;
    [self.contentView addSubview:_archiveButton];
    
    [NSLayoutConstraint activateConstraints:@[
        [_checkBox.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor constant:16],
        [_checkBox.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor],
        
        [_titleLabel.leadingAnchor constraintEqualToAnchor:_checkBox.trailingAnchor constant:12],
        [_titleLabel.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor],
        
        [_categoryLabel.leadingAnchor constraintEqualToAnchor:_titleLabel.trailingAnchor constant:8],
        [_categoryLabel.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor],
        
        [_archiveButton.leadingAnchor constraintEqualToAnchor:_categoryLabel.trailingAnchor constant:8],
        [_archiveButton.trailingAnchor constraintEqualToAnchor:self.contentView.trailingAnchor constant:-16],
        [_archiveButton.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor]
    ]];
}

- (void)configureWithTitle:(NSString *)title category:(NSString *)category isCompleted:(BOOL)isCompleted {
    _checkBox.on = isCompleted;
    _titleLabel.textColor = isCompleted ? [UIColor grayColor] : [UIColor blackColor];
    _categoryLabel.text = [NSString stringWithFormat:@"(%@)", category];
    _archiveButton.hidden = !isCompleted;
    
    if (isCompleted) {
        NSDictionary *attributes = @{NSStrikethroughStyleAttributeName: @(NSUnderlineStyleSingle)};
        _titleLabel.attributedText = [[NSAttributedString alloc] initWithString:title attributes:attributes];
    } else {
        _titleLabel.attributedText = nil;
        _titleLabel.text = title;
    }
}

- (void)checkChanged {
    BOOL isCompleted = _checkBox.on;
    _titleLabel.textColor = isCompleted ? [UIColor grayColor] : [UIColor blackColor];
    _archiveButton.hidden = !isCompleted;
    
    if (isCompleted) {
        NSDictionary *attributes = @{NSStrikethroughStyleAttributeName: @(NSUnderlineStyleSingle)};
        _titleLabel.attributedText = [[NSAttributedString alloc] initWithString:_titleLabel.text ?: @"" attributes:attributes];
    } else {
        _titleLabel.attributedText = nil;
        _titleLabel.text = _titleLabel.text;
    }
    
    if (self.onCheckChanged) {
        self.onCheckChanged(isCompleted);
    }
}

- (void)archiveTapped {
    if (self.onArchive) {
        self.onArchive();
    }
}

@end

File: _examples/swift-uikit/TodoViewController.swift
import UIKit

class TodoViewController: UIViewController {
    
    private let textField = UITextField()
    private let addButton = UIButton(type: .system)
    private let tableView = UITableView()
    private let categorySegment = UISegmentedControl(items: ["work", "daily"])
    
    private var todos: [(title: String, category: String, isCompleted: Bool)] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
    }
    
    private func setupUI() {
        view.backgroundColor = .white
        
        categorySegment.selectedSegmentIndex = 0
        categorySegment.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(categorySegment)
        
        textField.placeholder = "Enter todo"
        textField.borderStyle = .roundedRect
        textField.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(textField)
        
        addButton.setImage(UIImage(systemName: "plus.circle.fill"), for: .normal)
        addButton.translatesAutoresizingMaskIntoConstraints = false
        addButton.addTarget(self, action: #selector(addTodo), for: .touchUpInside)
        view.addSubview(addButton)
        
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.dataSource = self
        tableView.register(TodoCell.self, forCellReuseIdentifier: "TodoCell")
        view.addSubview(tableView)
        
        NSLayoutConstraint.activate([
            categorySegment.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16),
            categorySegment.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
            categorySegment.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
            
            textField.topAnchor.constraint(equalTo: categorySegment.bottomAnchor, constant: 16),
            textField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
            textField.trailingAnchor.constraint(equalTo: addButton.leadingAnchor, constant: -8),
            
            addButton.centerYAnchor.constraint(equalTo: textField.centerYAnchor),
            addButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
            addButton.widthAnchor.constraint(equalToConstant: 44),
            
            tableView.topAnchor.constraint(equalTo: textField.bottomAnchor, constant: 16),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
    }
    
    @objc private func addTodo() {
        guard let title = textField.text, !title.isEmpty else { return }
        let category = categorySegment.selectedSegmentIndex == 0 ? "work" : "daily"
        todos.append((title: title, category: category, isCompleted: false))
        textField.text = ""
        tableView.reloadData()
    }
}

extension TodoViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return todos.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "TodoCell", for: indexPath) as! TodoCell
        let todo = todos[indexPath.row]
        cell.configure(title: todo.title, category: todo.category, isCompleted: todo.isCompleted)
        cell.onCheckChanged = { [weak self] isCompleted in
            self?.todos[indexPath.row].isCompleted = isCompleted
        }
        cell.onArchive = { [weak self] in
            self?.todos.remove(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: .automatic)
        }
        return cell
    }
}

class TodoCell: UITableViewCell {
    private let checkBox = UISwitch()
    private let titleLabel = UILabel()
    private let categoryLabel = UILabel()
    private let archiveButton = UIButton(type: .system)
    var onCheckChanged: ((Bool) -> Void)?
    var onArchive: (() -> Void)?
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupUI() {
        checkBox.translatesAutoresizingMaskIntoConstraints = false
        checkBox.addTarget(self, action: #selector(checkChanged), for: .valueChanged)
        contentView.addSubview(checkBox)
        
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(titleLabel)
        
        categoryLabel.translatesAutoresizingMaskIntoConstraints = false
        categoryLabel.textColor = .gray
        contentView.addSubview(categoryLabel)
        
        archiveButton.translatesAutoresizingMaskIntoConstraints = false
        archiveButton.setTitle("Archive", for: .normal)
        archiveButton.addTarget(self, action: #selector(archiveTapped), for: .touchUpInside)
        archiveButton.isHidden = true
        contentView.addSubview(archiveButton)
        
        NSLayoutConstraint.activate([
            checkBox.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 16),
            checkBox.centerYAnchor.constraint(equalTo: contentView.centerYAnchor),
            
            titleLabel.leadingAnchor.constraint(equalTo: checkBox.trailingAnchor, constant: 12),
            titleLabel.centerYAnchor.constraint(equalTo: contentView.centerYAnchor),
            
            categoryLabel.leadingAnchor.constraint(equalTo: titleLabel.trailingAnchor, constant: 8),
            categoryLabel.centerYAnchor.constraint(equalTo: contentView.centerYAnchor),
            
            archiveButton.leadingAnchor.constraint(equalTo: categoryLabel.trailingAnchor, constant: 8),
            archiveButton.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -16),
            archiveButton.centerYAnchor.constraint(equalTo: contentView.centerYAnchor)
        ])
    }
    
    func configure(title: String, category: String, isCompleted: Bool) {
        titleLabel.text = title
        categoryLabel.text = "(\(category))"
        checkBox.isOn = isCompleted
        titleLabel.textColor = isCompleted ? .gray : .black
        
        if isCompleted {
            let attributedString = NSAttributedString(
                string: title,
                attributes: [.strikethroughStyle: NSUnderlineStyle.single.rawValue]
            )
            titleLabel.attributedText = attributedString
        } else {
            titleLabel.attributedText = nil
            titleLabel.text = title
        }
        
        archiveButton.isHidden = !isCompleted
    }
    
    @objc private func checkChanged() {
        let isCompleted = checkBox.isOn
        onCheckChanged?(isCompleted)
        
        titleLabel.textColor = isCompleted ? .gray : .black
        
        if isCompleted {
            let attributedString = NSAttributedString(
                string: titleLabel.text ?? "",
                attributes: [.strikethroughStyle: NSUnderlineStyle.single.rawValue]
            )
            titleLabel.attributedText = attributedString
        } else {
            titleLabel.attributedText = nil
            titleLabel.text = titleLabel.text
        }
        
        archiveButton.isHidden = !isCompleted
    }
    
    @objc private func archiveTapped() {
        onArchive?()
    }
}